raphtory/src/python/graph/history.rs

```diff
@@ -0,0 +1,793 @@
+        PyHistoryDateTime {
+            history_dt: HistoryDateTime::new(self.history.0.clone()), // clone the Arc, not the underlying object
+        }
+    }
+
+    /// Access secondary index (unique index) of history events
+    #[getter]
+    pub fn secondary_index(&self) -> PyHistorySecondaryIndex {
+        PyHistorySecondaryIndex {
+            history_s: HistorySecondaryIndex::new(self.history.0.clone()), // clone the Arc, not the underlying object
+        }
+    }
+
+    #[getter]
+    pub fn intervals(&self) -> PyIntervals {
+        PyIntervals {
+            intervals: Intervals::new(self.history.0.clone()), // clone the Arc, not the underlying object
+        }
+    }
+
+    /// Get the earliest time in the history
+    pub fn earliest_time(&self) -> Option<TimeIndexEntry> {
+        self.history.earliest_time()
+    }
+
+    /// Get the latest time in the history
+    pub fn latest_time(&self) -> Option<TimeIndexEntry> {
+        self.history.latest_time()
+    }
+
+    /// Collect all time events
+    pub fn collect(&self) -> Vec<TimeIndexEntry> {
+        self.history.collect()
+    }
+
+    /// Collect all time events in reverse
+    pub fn collect_rev(&self) -> Vec<TimeIndexEntry> {
+        self.history.collect_rev()
+    }
+
+    /// Iterate over all time events
+    pub fn __iter__(&self) -> PyBorrowingIterator {
+        py_borrowing_iter!(
+            self.history.clone(),
+            History<'static, Arc<dyn InternalHistoryOps>>,
+            |history| history.iter()
+        )
+    }
+
+    /// Iterate over all time events in reverse
+    pub fn __reversed__(&self) -> PyBorrowingIterator {
+        py_borrowing_iter!(
+            self.history.clone(),
+            History<'static, Arc<dyn InternalHistoryOps>>,
+            |history| history.iter_rev()
+        )
+    }
+
+    pub fn __repr__(&self) -> String {
+        self.history.repr()
+    }
+
+    fn __hash__(&self) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        self.history.hash(&mut hasher);
+        hasher.finish()
+    }
+
+    fn __contains__(&self, item: TimeIndexEntry) -> bool {
+        self.history.iter().any(|x| x == item)
+    }
+
+    fn __eq__(&self, other: &Bound<PyAny>) -> bool {
+        if let Ok(py_hist) = other.downcast::<PyHistory>() {
+            return self.history.eq(&py_hist.get().history);
+        }
+        if let Ok(list) = other.extract::<Vec<TimeIndexEntry>>() {
+            return self.history.iter().eq(list.into_iter());
+        }
+        false
+    }
+
+    fn __ne__(&self, other: &Bound<PyAny>) -> bool {
+        !self.__eq__(other)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.history.is_empty()
+    }
+
+    pub fn __len__(&self) -> usize {
+        self.history.len()
+    }
+}
+
+impl<T: InternalHistoryOps + 'static> From<History<'_, T>> for PyHistory {
+    fn from(history: History<T>) -> Self {
+        let arc_ops: Arc<dyn InternalHistoryOps> = {
```

Comment:
might be easier and cleaner to add an `as_arc_dyn` method to InternalHistoryOps which you can implement appropriately? 

with this default implementation:
```rust
fn as_arc_dyn(&self) -> Arc<dyn InternalHistoryOps> {
    Arc::new(self.clone())
}
```

and for `Arc<dyn InternalHistoryOps>` you can have 
```rust
fn as_arc_dyn(&self) -> Arc<dyn InternalHistoryOps> {
    self.clone()
}
```